torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -820,13 +938,83 @@ def auto_functionalized_v2_fake(
+        # kwargs may contain general callables, that are not proxable e.g. FunctionWithNoFreeVars
+        # this could happen when we auto_functionalize the backward of the hop,
+        # where backward fn is a callablle that wrapps forward graph module.
+        # This function materialize the callable args according to the schema of the hop.
+
+        # We cannot materialize the callables in kwargs directly because the inputs to callable
+        # vary from hops to hop. To make the materialiation process generic to all hops,
+        # we trace a function that wraps the hop and let each hop itself figure out how to trace
+        # its callable inputs. Then we look at the schema of the traced hop node and replace the
+        # callable in original kwarg with the traced subgraphs.
+        #
+        # Specifically, we first trace a wrapped_fn that calls into the hop. Then we look for the
+        # hop node in the traced graph and graph module inputs to the hop. Finally, we replace the
+        # original kwarg's callable with the graph module.
+        all_bases = kwargs.get("_all_bases", [])
+        _only_clone_these_bases = kwargs.get("_only_clone_these_bases", None)
+        if _only_clone_these_bases is None:
+            _only_clone_these_bases = tuple(range(len(all_bases)))
+
+        schema = pytree.tree_unflatten([], kwargs.get("_op_schema", None)).schema
+        new_kwargs, _ = _generate_new_op_kwargs_from_bases(
+            schema,
+            {k: v for k, v in kwargs.items() if k not in ("_all_bases", "_op_schema")},
+            all_bases,
+            _only_clone_these_bases,
+        )
+
+        def wrapped_fn(*args):
+            return _invoke_op_with_kwargs_and_schema(_mutable_op, new_kwargs, schema)  # type: ignore[arg-type]
+
+        # We need to trace the higher order op in order to materilaize the callable inputs that
+        # are a callable (e.g. after functionalization key)
+        gm = reenter_make_fx(wrapped_fn)(pytree.tree_leaves(new_kwargs))
+        hop_node = gm.graph.find_nodes(op="call_function", target=_mutable_op)[0]
+        arg_proxies = pytree.tree_leaves((hop_node.args, hop_node.kwargs))
+        assert isinstance(schema, torch._C.FunctionSchema) and len(arg_proxies) == len(
+            schema.arguments
+        )
+
+        # _invoke_op_with_kwargs_and_schema preserves ordering of proxies via schema
+        materialized_arg = {}
+        for proxy, arg in zip(arg_proxies, schema.arguments):
+            if (
+                isinstance(proxy, torch.fx.Node)
+                and proxy.op == "get_attr"
+                and isinstance(getattr(gm, proxy.target), torch.fx.GraphModule)  # type: ignore[arg-type]
+            ):
+                assert callable(kwargs[arg.name]), (schema, arg.name, kwargs)
+                materialized_arg[arg.name] = getattr(gm, proxy.target)  # type: ignore[arg-type]
+
+        # Update kwargs with materialized graphs
+        kwargs.update(materialized_arg)
```

Comment:
Can you shove this into a helper function? Given a HOP, some args, and some kwargs, it would return the materialized graphs

Reply:
Can also go into higher_order_ops.utils